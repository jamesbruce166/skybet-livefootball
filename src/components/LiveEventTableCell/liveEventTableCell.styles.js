import styled from 'styled-components';
import { ChevronRight } from '@styled-icons/bootstrap';
import { SECONDARY_DARK } from '../../variables.styles';

export const CellContent = styled.div`
	display: flex;
	flex-direction: row;
	flex-wrap: wrap;
	justify-content: center;
	align-items: center;
	padding: 10px 25px;
	background: ${SECONDARY_DARK};
	margin-top: 12px;
	border-radius: 8px;
`;

export const CompetitorBox = styled.div`
	display: flex;
	flex-direction: column;
	width: 60%;
`;

export const SubRow = styled.div`
	display: flex;
	flex-direction: row;
	flex-wrap: wrap;
	justify-content: space-between;
`;

export const TeamNameText = styled.p`
	color: #ffffff;
	font-size: 0.9rem;
	font-weight: 300;
	line-height: 0;
	letter-spacing: 1px;
`;

export const ScoreText = styled.p`
	color: rgb(255, 30, 60);
	font-size: 0.9rem;
	font-weight: 600;
	line-height: 0;
`;

export const StatusBox = styled.div`
	display: flex;
	flex-direction: row;
	justify-content: flex-end;
	align-items: center;
	width: 39%;
	gap: 15px;
`;

export const Chevron = styled(ChevronRight)`
	color: lightgray;
	width: 18px;
`;
