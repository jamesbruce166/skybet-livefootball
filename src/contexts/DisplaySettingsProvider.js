import React, { useContext, useState } from 'react';

export const DisplaySettingsContext = React.createContext();

export function useDisplaySettings() {
	return useContext(DisplaySettingsContext);
}

export function DisplaySettingsProvider({ children }) {
	const displayOptions = {
		FRACTION: 'decimal',
		DECIMAL: 'fraction',
	};
	const [oddsDisplay, setOddsDisplay] = useState(displayOptions.DECIMAL);

	const value = {
		oddsDisplay,
		setOddsDisplay,
		displayOptions,
	};

	return (
		<DisplaySettingsContext.Provider value={value}>
			{children}
		</DisplaySettingsContext.Provider>
	);
}
