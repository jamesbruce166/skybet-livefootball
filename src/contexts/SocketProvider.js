import React, { useContext, useEffect, useState, useRef } from 'react';

const SocketContext = React.createContext();

export function useSocket() {
	return useContext(SocketContext);
}

export function SocketProvider({ id, children }) {
	const [socket, setSocket] = useState();
	const ws = useRef(null);

	useEffect(() => {
		ws.current = new WebSocket(process.env.REACT_APP_SOCKET_URL);

		ws.current.addEventListener('open', () => {
			ws.current.send(
				JSON.stringify({
					type: 'getLiveEvents',
					primaryMarkets: true,
				})
			);
		});
		setSocket(ws.current);
	}, [id]);

	const value = {
		socket,
	};

	return (
		<SocketContext.Provider value={value}>
			{children}
		</SocketContext.Provider>
	);
}
