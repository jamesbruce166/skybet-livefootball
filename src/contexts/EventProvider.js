import React, {
	useContext,
	useState,
	useCallback,
	createContext,
	useEffect,
} from 'react';
import { useErrorHandler } from 'react-error-boundary';

import { useSocket } from './SocketProvider';

export const EventsContext = createContext();

export function useEvents() {
	return useContext(EventsContext);
}

export function EventProvider({ children }) {
	const [selectedEvent, setSelectedEvent] = useState(undefined);
	const [events, setEvents] = useState(undefined);
	const handleError = useErrorHandler();

	const { socket } = useSocket();

	const onMessage = useCallback((m) => {
		const data = JSON.parse(m.data);
		const { type } = data;
		switch (type) {
			case 'LIVE_EVENTS_DATA':
				const firstGame = data.data && data.data[0];

				setEvents(data);
				setSelectedEvent(firstGame);
				break;
			case 'ERROR':
				handleError(data);
				break;
			default:
				break;
		}
	}, []);

	useEffect(() => {
		if (socket == null) return;

		socket.addEventListener('message', onMessage);

		return () => {
			socket.removeEventListener('message', onMessage);
		};
	}, [socket, onMessage]);

	const value = {
		events,
		selectedEvent,
		setSelectedEvent,
	};

	return (
		<EventsContext.Provider value={value}>
			{children}
		</EventsContext.Provider>
	);
}
