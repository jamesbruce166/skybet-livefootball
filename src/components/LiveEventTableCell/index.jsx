import React from 'react';
import PropTypes from 'prop-types';

import {
	CellContent,
	CompetitorBox,
	SubRow,
	TeamNameText,
	ScoreText,
	StatusBox,
	Chevron,
} from './liveEventTableCell.styles';
import { LiveBadge } from '../common/common.styles';

import { useEvents } from '../../contexts/EventProvider';

const LiveEventTableCell = ({ event }) => {
	const { competitors, scores, status } = event;
	const { setSelectedEvent } = useEvents();

	const handleClickEvent = () => setSelectedEvent(event);

	const { name: homeTeamName } = competitors[0];
	const { name: awayTeamName } = competitors[1];
	const { home, away } = scores;
	const { live } = status;

	return (
		<CellContent data-testid='event-cell' onClick={handleClickEvent}>
			<CompetitorBox>
				<SubRow>
					<TeamNameText>{homeTeamName}</TeamNameText>
					<ScoreText>{home}</ScoreText>
				</SubRow>
				<SubRow>
					<TeamNameText>{awayTeamName}</TeamNameText>
					<ScoreText>{away}</ScoreText>
				</SubRow>
			</CompetitorBox>
			<StatusBox>
				{live && <LiveBadge>LIVE</LiveBadge>}
				<Chevron />
			</StatusBox>
		</CellContent>
	);
};

LiveEventTableCell.propTypes = {
	event: PropTypes.object.isRequired,
};

export default LiveEventTableCell;
