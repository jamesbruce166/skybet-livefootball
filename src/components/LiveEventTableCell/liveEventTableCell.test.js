import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, cleanup } from '@testing-library/react';

import { EventsContext } from '../../contexts/EventProvider';
import { SocketContext } from '../../contexts/SocketProvider';
import { DisplaySettingsContext } from '../../contexts/DisplaySettingsProvider';

import LiveEventTableCell from './index';
import mockEvent from './__mocks__/liveEvent.json';

afterEach(() => {
	cleanup();
});

function renderLiveEventTableCell(event) {
	const displayOptions = {
		FRACTION: 'decimal',
		DECIMAL: 'fraction',
	};
	const oddsDisplay = displayOptions.DECIMAL;
	return (
		<DisplaySettingsContext.Provider
			value={{ oddsDisplay, displayOptions }}
		>
			<SocketContext.Provider value={{}}>
				<EventsContext.Provider value={{ setSelectedEvent: () => {} }}>
					<LiveEventTableCell event={event} />
				</EventsContext.Provider>
			</SocketContext.Provider>
		</DisplaySettingsContext.Provider>
	);
}

test('renders correctly', () => {
	const tree = renderer.create(renderLiveEventTableCell(mockEvent)).toJSON();
	expect(tree).toMatchSnapshot();
});

test('should display home team name', () => {
	render(renderLiveEventTableCell(mockEvent));
	const eventCell = screen.queryByTestId('event-cell');

	expect(eventCell).toBeInTheDocument();
	expect(eventCell).toHaveTextContent(mockEvent.competitors[0].name);
});

test('should display away team name', () => {
	render(renderLiveEventTableCell(mockEvent));
	const eventCell = screen.queryByTestId('event-cell');

	expect(eventCell).toBeInTheDocument();
	expect(eventCell).toHaveTextContent(mockEvent.competitors[1].name);
});

test('should display home team score', () => {
	render(renderLiveEventTableCell(mockEvent));
	const eventCell = screen.queryByTestId('event-cell');

	expect(eventCell).toBeInTheDocument();
	expect(eventCell).toHaveTextContent(mockEvent.scores.home);
});

test('should display away team score', () => {
	render(renderLiveEventTableCell(mockEvent));
	const eventCell = screen.queryByTestId('event-cell');

	expect(eventCell).toBeInTheDocument();
	expect(eventCell).toHaveTextContent(mockEvent.scores.away);
});

test('should display is live', () => {
	render(renderLiveEventTableCell(mockEvent));
	const eventCell = screen.queryByTestId('event-cell');

	expect(eventCell).toBeInTheDocument();
	expect(eventCell).toHaveTextContent('LIVE');
});

test('should hide live status when finished', () => {
	mockEvent.status.live = false;
	render(renderLiveEventTableCell(mockEvent));

	mockEvent.status.live = true;
	const eventCell = screen.queryByTestId('event-cell');

	expect(eventCell).toBeInTheDocument();
	expect(eventCell).not.toHaveTextContent('LIVE');
});
