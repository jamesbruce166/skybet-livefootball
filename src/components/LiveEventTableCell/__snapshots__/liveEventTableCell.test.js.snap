// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bczRLJ cqpXMw"
  data-testid="event-cell"
  onClick={[Function]}
>
  <div
    className="sc-gsnTZi jfffIx"
  >
    <div
      className="sc-dkzDqf caZALy"
    >
      <p
        className="sc-hKMtZM bcXSRW"
      >
        Shanghai Shenhua
      </p>
      <p
        className="sc-eCYdqJ inrNUh"
      >
        0
      </p>
    </div>
    <div
      className="sc-dkzDqf caZALy"
    >
      <p
        className="sc-hKMtZM bcXSRW"
      >
        Shandong Luneng Taishan
      </p>
      <p
        className="sc-eCYdqJ inrNUh"
      >
        0
      </p>
    </div>
  </div>
  <div
    className="sc-jSMfEi fnIhHY"
  >
    <div
      className="sc-iBkjds gvCYMO"
    >
      LIVE
    </div>
    <svg
      aria-hidden="true"
      className="StyledIconBase-ea9ulj-0 lbJwfL sc-gKXOVf clsDmZ"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4.646 1.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1 0 .708l-6 6a.5.5 0 0 1-.708-.708L10.293 8 4.646 2.354a.5.5 0 0 1 0-.708z"
        fillRule="evenodd"
      />
    </svg>
  </div>
</div>
`;
